inseq/attr/feat/ops/basic_attention.py

```diff
@@ -0,0 +1,247 @@
+                "decoder_attention_mask": forward_args[5] if is_target_attr else forward_args[6],
+            }
+
+            if use_embeddings:
+                forward_pass_args["inputs_embeds"] = inputs[0]
+                forward_pass_args["decoder_inputs_embeds"] = inputs[1] if is_target_attr else forward_args[0]
+            else:
+                forward_pass_args["input_ids"] = forward_args[0] if is_target_attr else forward_args[1]
+                forward_pass_args["decoder_input_ids"] = forward_args[1] if is_target_attr else forward_args[2]
+
+        else:
+
+            use_embeddings = forward_args[4]
+
+            forward_pass_args = {"attention_mask": forward_args[3]}
+
+            if use_embeddings:
+                forward_pass_args["inputs_embeds"] = inputs[0]
+            else:
+                forward_pass_args["input_ids"] = forward_args[0]
+
+        return forward_pass_args
+
+
+class AggregatedAttention(AttentionAttribution):
+    """
+    A basic attention attribution approach.
+    It will return the attention values averaged across all layers.
+    """
+
+    @log_usage()
+    def attribute(
+        self,
+        inputs: MultiStepEmbeddingsTensor,
+        target: TargetType = None,
+        merge_head_option: str = "average",
+        use_head: int = None,
+        additional_forward_args: Any = None,
+    ) -> Union[TensorOrTupleOfTensorsGeneric, Tuple[TensorOrTupleOfTensorsGeneric, torch.Tensor]]:
+
+        is_inputs_tuple = _is_tuple(inputs)
+
+        is_target_attribution = True if len(inputs) > 1 else False
+
+        is_encoder_decoder = self.forward_func.is_encoder_decoder
+
+        forward_pass_args = self._extract_forward_pass_args(
+            inputs, additional_forward_args, is_target_attribution, is_encoder_decoder
+        )
+
+        outputs = self.forward_func.model(**forward_pass_args)
+
+        if is_encoder_decoder:
+            cross_aggregation = torch.stack(outputs.cross_attentions).mean(0)
+            cross_aggregation = self._merge_attention_heads(cross_aggregation, merge_head_option, use_head)
+            cross_aggregation = cross_aggregation.select(1, -1)
+
+            attributions = (cross_aggregation,)
+
+            if is_target_attribution:
+                decoder_aggregation = torch.stack(outputs.decoder_attentions).mean(0)
+                decoder_aggregation = self._merge_attention_heads(decoder_aggregation, merge_head_option, use_head)
+                decoder_aggregation = decoder_aggregation.select(1, -1)
+
+                attributions = attributions + (decoder_aggregation,)
+        else:
+            aggregation = torch.stack(outputs.attentions).mean(0)
+            aggregation = self._merge_attention_heads(aggregation, merge_head_option, use_head)
+            aggregation = aggregation.select(1, -1)
+
+            attributions = (aggregation,)
+
+        return _format_output(is_inputs_tuple, attributions)
+
+
+class LastLayerAttention(AttentionAttribution):
```

Comment:
What is the benefit of two separate classes for basic attention attribution? Wouldn't a single class in which `layer`, `head`, `aggregate_head_fn` and `aggregate_layer_fn` can be customized as parameters of `model.attribute` make more sense?

Reply:
true, I think we also talked about that shortly before Christmas. I would argue leaving them in for usage simplicity but add a third class which allows for the configurations you mention and refactor the other two classes to inherit from this class with the specific parameter values preset already for their use cases.

Reply:
Yes, this seems like a reasonable choice to me!

Reply:
When taking a closer look just now I realized that merging the two classes from a common subclass does not make a lot of sense, since there are always two different functions needed dependent on if we use a single layer or a range of layers. So I simply renamed `LastLayerAttention` to `SingleLayerAttention` and added a parameter to choose a specific layer which will default to the last layer when not set. Similar for `AggregatedAttention` I added a parameter to define a range of layers to average over.